src/FluidStructureProblem.hpp

```diff
@@ -0,0 +1,693 @@
+                               src.block(0),
+                               *preconditioner_velocity);
+#ifdef DEBUG
+      std::cout << "  " << solver_control_velocity.last_step()
+                << " CG1 iterations" << std::endl;
+#endif
+      tmpStokes.reinit(src.block(1));
+      B->vmult(tmpStokes, dst.block(0));
+      tmpStokes.sadd(-1.0, src.block(1));
+
+      SolverControl solver_control_pressure(1000, 1e-2 * tmpStokes.l2_norm());
+      SolverCG<TrilinosWrappers::MPI::Vector> solver_cg_pressure(
+        solver_control_pressure);
+      solver_cg_pressure.solve(*pressure_mass,
+                               dst.block(1),
+                               tmpStokes,
+                               *preconditioner_pressure);
+#ifdef DEBUG
+      std::cout << "  " << solver_control_pressure.last_step()
+                << " CG2 iterations" << std::endl;
+#endif
+      tmpStokes.reinit(src.block(2));
+      D1->vmult(tmpStokes, dst.block(0));
+      D2->vmult_add(tmpStokes, dst.block(1));
+      tmpStokes.sadd(-1.0, src.block(2));
+
+
+      // other way to do it, after testing we found it to be extremely slow.
+      // preconditioner_solid.vmult(dst.block(2), tmpStokes);
+
+      SolverControl solver_control_solid(1000, 1e-2 * tmpStokes.l2_norm());
+      SolverCG<TrilinosWrappers::MPI::Vector> solver_cg_solid(
+        solver_control_solid);
+
+      solver_cg_solid.solve(*solid_matrix,
+                            dst.block(2),
+                            tmpStokes,
+                            *preconditioner_solid);
+#ifdef DEBUG
+      std::cout << "  " << solver_control_solid.last_step() << " CG3 iterations"
+                << std::endl;
+#endif
+    }
+
+  protected:
+    // Velocity stiffness matrix.
+    const LA::MPI::SparseMatrix *velocity_stiffness;
+
+    // Preconditioner used for the velocity block.
+    std::shared_ptr<TrilinosWrappers::PreconditionAMG> preconditioner_velocity;
+
+    // Pressure mass matrix.
+    const LA::MPI::SparseMatrix *pressure_mass;
+
+    // Preconditioner used for the pressure block.
+    std::shared_ptr<TrilinosWrappers::PreconditionAMG> preconditioner_pressure;
+
+    // B matrix.
+    const LA::MPI::SparseMatrix *B;
+
+    // D1 matrix.
+    const LA::MPI::SparseMatrix *D1;
+
+    // D2 matrix.
+    const LA::MPI::SparseMatrix *D2;
+
+    // Preconditioner used for the pressure block.
+    std::shared_ptr<TrilinosWrappers::PreconditionAMG> preconditioner_solid;
+
+    // Solid matrix.
+    const LA::MPI::SparseMatrix *solid_matrix;
+
+    // Temporary vector stokes
+    mutable LA::MPI::Vector tmpStokes;
+  };
+
+  // Searching for other preconditioner to implement, we also tried a sort of
+  // SIMPLE preconditioner used for Navier-Stokes equations. However, the
+  // performance was not satisfactory, so we decided not to use it. The code is
+  // left here commented for possible future reference. class
+  // PreconditionBlockTriangularSimple
```

Comment:
This is interesting: why didn't you report the performance of this preconditioner in the report?